hub/core/chunk_engine/write.py

```diff
@@ -29,43 +40,103 @@ def write_array(
         batched (bool): If True, the provied `array`'s first axis (`shape[0]`) will be considered it's batch axis.
             If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
-        tobytes (Callable): Callable that flattens an array into bytes. Must accept an `np.ndarray` as it's argument and return `bytes`.
 
     Raises:
-        NotImplementedError: Do not use this function for writing to a key that already exists.
+        KeyAlreadyExistsError: If trying to write to a tensor that already exists in `storage` under `key`.
     """
 
-    array = normalize_and_batchify_shape(array, batched=batched)
+    if key_exists(key, storage):
+        raise KeyAlreadyExistsError(key, "Use `append_array`.")
 
-    meta_key = get_meta_key(key)
-    index_map_key = get_index_map_key(key)
-    if meta_key in storage or index_map_key in storage:
-        # TODO: when appending is done change this error
-        raise NotImplementedError("Appending is not supported yet.")
+    array = normalize_and_batchify_shape(array, batched=batched)
 
     index_map: List[dict] = []
     meta = {
         "chunk_size": chunk_size,
         "dtype": array.dtype.name,
         "length": array.shape[0],
-        "min_shape": array.shape[1:],
-        "max_shape": array.shape[1:],
+        "min_shape": tuple(array.shape[1:]),
+        "max_shape": tuple(array.shape[1:]),
+        # TODO: add entry in meta for which tobytes function is used and handle mismatch versions for this
     }
 
+    write_samples(array, key, storage, meta, index_map)
+
+    write_tensor_meta(key, storage, meta)
+    write_index_map(key, storage, index_map)
+
+
+def append_array(
+    array: np.ndarray,
+    key: str,
+    storage: StorageProvider,
+    batched: bool = False,
+):
+    """Chunk and write the given array to an already existing tensor in storage. For writing an array to a tensor that does not exist, use `write_array`.
+
+    For more on chunking, see the `generate_chunks` method.
+
+    Args:
+        array (np.ndarray): Array to be chunked/written. Batch axis (`array.shape[0]`) is optional, if `array` does have a
+            batch axis, you should pass the argument `batched=True`.
+        key (str): Key for where the chunks, index_map, and meta are located in `storage` relative to it's root.
+        storage (StorageProvider): StorageProvider where the chunks, index_map, and meta are stored.
+        batched (bool): If True, the provied `array`'s first axis (`shape[0]`) will be considered it's batch axis.
+            If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
+
+    Raises:
+        KeyDoesNotExistError: If trying to append to a tensor that does not exist in `storage` under `key`.
+    """
+
+    if not key_exists(key, storage):
+        raise KeyDoesNotExistError(key, "Use `write_array`.")
+
+    array = normalize_and_batchify_shape(array, batched=batched)
```

Comment:
Are we ever denormalizing the shape? I feel like we should store the shape beforehand and do a reshape on read so we don't give them a shape they aren't expecting

Reply:
i don't think we should store the original shape, often the storage of arrays in a denormalized manner are not out of preference, but out of circumstances

Reply:
if it becomes a common complaint though, we can definitely change it. i don't see a need to do so now though